commons/src/main/java/fr/free/nrw/commons/caching/QuadTree.java

```diff
@@ -0,0 +1,477 @@
+package fr.free.nrw.commons.caching;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * Datastructure: A point Quad Tree for representing 2D data. Each
+ * region has the same ratio as the bounds for the tree.
+ * <p/>
+ * The implementation currently requires pre-determined bounds for data as it
+ * can not rebalance itself to that degree.
+ */
+public class QuadTree {
```

Comment:
https://github.com/varunpant/Quadtree does not provide official JAR releases unfortunately. We could fork Quadtree and release a JAR, that would simplify our project's source code indeed.